sw/device/bsp/w25q/w25q.c

```diff
@@ -0,0 +1,1247 @@
+
+/**
+ * @brief Write (up to) a page to the flash.
+ * 
+ * @param addr 24-bit address to write to.
+ * @param data pointer to the data buffer.
+ * @param length number of bytes to write.
+ * @param quad if 1, the write is performed at quad speed.
+ * @param dma if 1, the write is performed using DMA.
+*/
+static void page_write(uint32_t addr, uint8_t *data, uint32_t length, uint8_t quad, uint8_t dma);
+
+/**
+ * @brief Copy length bytes from data to the SPI TX FIFO, using DMA.
+ * 
+ * @param data pointer to the data buffer.
+ * @param length number of bytes to copy.
+*/
+static uint8_t dma_send_toflash(uint8_t *data, uint32_t length);
+
+/**
+ * @brief Enable flash write.
+ * 
+ * It sets the WEL bit in the flash status register.
+ * Every action that require the WEL to be set is goig to automatically
+ * clear it.
+*/
+static void flash_write_enable(void);
+
+/**
+ * @brief Performs sanity checks on the input parameters.
+ * 
+ * Checks if the address is valid, the data pointer is not NULL
+ * and the length is not 0.
+ * 
+ * @param addr 24-bit address.
+ * @param data pointer to the data buffer.
+ * @param length number of bytes to read/write.
+ * @return 1 if the checks are passed, 0 otherwise.
+*/
+static uint8_t sanity_checks(uint32_t addr, uint8_t *data, uint32_t length);
+
+/**
+ * @brief Return the minimum between two numbers.
+ * 
+ * The function uses signed integers in order to handle also negative numbers.
+ * 
+ * @param a first number.
+ * @param b second number.
+ * @return the minimum between a and b.
+*/
+static int32_t MIN(int32_t a, int32_t b) {
+    return (a < b) ? a : b;
+}
+
+
+/****************************************************************************/
+/**                                                                        **/
+/*                            GLOBAL VARIABLES                              */
+/**                                                                        **/
+/****************************************************************************/
+
+/**
+ * @brief SPI structure.
+*/
+spi_host_t spi;
+
+/**
+ * @brief Static vector used in the erase_and_write function.
+ * 
+ * It is used to store the data of a sector (4k) of the flash. Given the dimensions
+ * of the vector, is not possible to allocate it dinamically as it would cause a stack
+ * or heap overflow.
+*/
+uint8_t sector_data[4096];
+
+
+/****************************************************************************/
+/**                                                                        **/
+/*                           EXPORTED FUNCTIONS                             */
+/**                                                                        **/
+/****************************************************************************/
+uint8_t w25q128jw_init(spi_host_t spi_host) {
+    soc_ctrl_t soc_ctrl;
+    soc_ctrl.base_addr = mmio_region_from_addr((uintptr_t)SOC_CTRL_START_ADDRESS);
```

Comment:
you do not need to do this, you already have the BA in here --> `soc_ctrl_peri` that's a macro that is automatically generated after mcu-gen in the file `soc_ctrl_strcuts.h`

Reply:
In the struct there is not the base address, but there is already the information that I needed (SPIMEMIO yes or no) so I guess is even better now.

Reply:
the BA is not in the struct, but in the define of the structs file

Reply:
Yes, but is casted to a different type name (`soc_ctrl` != `soc_ctrl_t`) so I should have done a cast making it actually more complicated to read.
If you prefer I can revert the change and use `soc_ctrl_peri` as a sort of base address.